### ISSUE TITLE
equals() method does a type checking while comparing?

### ISSUE THREAD

**Issue by kamleshchandnani** [dfahlander/Dexie.js#515]
Sample data:
`users =  { id:2, name:'kamlesh' }`
`Dexie.users.where('id').equals('2').toArray()`
returns empty array.
but 
`Dexie.users.where('id').equals(2).toArray()` returns the data

**Comment by nponiros** [comment]
I don't think that there is an explicit type check here. In your first example you compare string to number and I'm guessing dexie uses strict equality (===) here so `2===2` is false.

**Comment by kamleshchandnani** [comment]
@nicolashenry Yup that is what I wanted to confirm. So is there any other alternative to equals where strict equality is not a constraint?

**Comment by nponiros** [comment]
None that I know off but I like strict equality. You could use each () on the table and filter on your own.

**Comment by kamleshchandnani** [comment]
Isn't using each on all the records costly compared to equals?

**Comment by nponiros** [comment]
I don't remember how equals is implemented but you are right that each is probably more costly due to the callback call but I doubt that you will notice a difference. It depends on how much data you have and on the device you are implementing for. You could try it and see if it is too slow.

**Comment by dfahlander** [comment]
If you really want a fuzzy search, i think the most performant query would be:
```js
db.users.where('id').anyOf(2, '2').toArray();
```

**Comment by kamleshchandnani** [comment]
@dfahlander Yup but again I have to add explicit anyOf
Like I have dynamic queries where ids come from 3rd party APIs and these api's gives numeric ids which are in the form '2' or 2 so each time I have to do
`db.users.where('id').anyOf(parseInt(prams.id), params.id).toArray();`

**Comment by kamleshchandnani** [comment]
@dfahlander makes a case when I do orderby and if data(id) is stored as string then I have to even write my own sort method and iterate over it

**Comment by dfahlander** [comment]
@kamleshchandnani This is how indexedDB works. If you may have mixed types in your existing customer's data, you could fix this by defining a new database version with an upgrader that canonicalizes it:

```js
db.version(2).upgrade(async ()=> {
    // Get all string-id types:
    let friendsWithWrongType = await db.friends.where('id').between('0', 'A').toArray();
    await db.friends.bulkDelete(friendsWithWrongType.map(f => f.id));
    let friendsWithCorrectedType = friendsWithWrongType.map(f => {...f, id: Number(f.id)});
    await db.friends.bulkPut(friendsWithCorrectedType);
});
```
